src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,449 @@
+                foreach (var closure in scope.Closures)
+                {
+                    action(scope, closure);
+                }
+                foreach (var nested in scope.NestedScopes)
+                {
+                    VisitClosures(nested, action);
+                }
+            }
+
+            private class ScopeTreeBuilder : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+            {
+                private Scope _currentScope;
+                // Null if we're not inside a closure scope, otherwise the nearest closure scope
+                private Closure _currentClosure;
+                private bool _inExpressionTree = false;
+
+                // A mapping from all captured vars to the scope they were declared in. This
+                // is used when recording captured variables as we must know what the lifetime
+                // of a captured variable is to determine the lifetime of its capture environment.
+                private readonly SmallDictionary<Symbol, Scope> _localToScope = new SmallDictionary<Symbol, Scope>();
+
+                private readonly Analysis _analysis;
+
+                private ScopeTreeBuilder(Scope rootScope, Analysis analysis)
+                {
+                    _currentScope = rootScope;
+                    _analysis = analysis;
+                    _currentClosure = null;
+                }
+
+                public static Scope Build(BoundNode node, Analysis analysis)
+                {
+                    // This should be the top-level node
+                    Debug.Assert(node == FindNodeToAnalyze(node));
+
+                    var rootScope = new Scope(parent: null, boundNode: node, containingClosure: null);
+                    var builder = new ScopeTreeBuilder(rootScope, analysis);
+                    builder.Build();
+                    return rootScope;
+                }
+
+                private void Build()
+                {
+                    // Set up the current method locals
+                    DeclareLocals(_currentScope, _analysis._topLevelMethod.Parameters);
+                    Visit(_currentScope.BoundNode);
+                }
+
+                public override BoundNode VisitMethodGroup(BoundMethodGroup node)
+                    => throw ExceptionUtilities.Unreachable;
+
+                public override BoundNode VisitBlock(BoundBlock node)
+                {
+                    if (node.Locals.IsDefaultOrEmpty)
+                    {
+                        // Skip introducing a new scope if there are no new locals
+                        return base.VisitBlock(node);
+                    }
+
+                    var oldScope = _currentScope;
+                    _currentScope = NewNestedScope(node);
+                    DeclareLocals(_currentScope, node.Locals);
+                    var result = base.VisitBlock(node);
+                    _currentScope = oldScope;
+                    return result; 
```

Comment:
These 6 lines are repeated a number of times. Consider moving into a common method.